os/drivers/input/ist415.c

```diff
@@ -0,0 +1,407 @@
+	ist415_close,  /* close */
+	ist415_read,   /* read */
+	ist415_write,           /* write */
+	NULL,           /* seek */
+	NULL,          /*ioctl */
+	ist415_poll,          
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static int ist415_get_touch_data(struct ist415_dev_s *dev, FAR void *buf)
+{
+	struct i2c_dev_s *i2c = dev->i2c;
+	struct i2c_config_s config = dev->config;
+	struct ts_event_coordinate *p_evt_coord;
+	struct touch_sample_s *data = buf;
+	u8 event[EVENT_PACKET_SIZE];
+	u8 *left_event;
+	int reg[2] = {HCOM_GET_EVENT_1ST, 0};
+	int ret = i2c_write(i2c, &config, (uint8_t *)reg, 1);
+	if (ret == 1) {
+		i2c_read(i2c, &config, (uint8_t *)event, EVENT_PACKET_SIZE);
+		u8 le  = event[7] & 0x1F;
+		u8 eid = event[0] & 0x3;
+		if (eid == EID_COORD) {
+			p_evt_coord = (struct ts_event_coordinate *) event;
+			data->npoints = 1;
+  			data->point[0].id = p_evt_coord->tid - 1;
+  			data->point[0].x  = (p_evt_coord->x_11_4 << 4) | p_evt_coord->x_3_0;
+  			data->point[0].y  = (p_evt_coord->y_11_4 << 4) | p_evt_coord->y_3_0;
+  			data->point[0].h = 0;
+  			data->point[0].w = 0;
+  			data->point[0].pressure = 0;
+			switch (p_evt_coord->tsta) {
+			case 1:
+				data->point[0].flags = TOUCH_DOWN;
+				break;
+			case 2:
+				data->point[0].flags = TOUCH_MOVE;
+				break;
+			case 3:
+				data->point[0].flags = TOUCH_UP;
+				break;
+			}
+			touchvdbg("coordinates are id %d status %d type %d x : %d y : %d\n", data->point[0].id, p_evt_coord->tsta, (p_evt_coord->ttype_3_2 << 2) | p_evt_coord->ttype_1_0, data->point[0].x, data->point[0].y);
+			return 1;
+		} else if (eid == EID_STATUS) {
+			/** TO DO **/	
+		}
+	}
+	return -1;
+}
+
+/************************************************************************************
+ * Name: touch_semtake
+ ************************************************************************************/
+
+static inline int touch_semtake(FAR sem_t *sem, bool errout)
+{
+	/* Take a count from the semaphore, possibly waiting */
+	if (sem_wait(sem) < 0) {
+		/* EINTR is the only error that we expect */
+		int errcode = get_errno();
+		DEBUGASSERT(errcode == EINTR);
+		if (errout) {
+			return errcode;
+		}
+	}
+	return 0;
+}
+
+/****************************************************************************
+ * Name: touch_semgive
+ ****************************************************************************/
+
+void touch_semgive(sem_t *sem)
+{
+	sem_post(sem);
+}
+
+static int ist415_open(FAR struct file *filep)
+{
+	FAR struct ist415_dev_s *priv;
+	priv = filep->f_inode->i_private;
+
+	if (!priv) {
+		return -EINVAL;
+	}
+
+	touch_semtake(&priv->sem, false);
+	if (priv->crefs == 0) {
+		priv->lower->irq_enable();
+	}
+	priv->crefs++;
+	DEBUGASSERT(priv->crefs > 0);
+	touch_semgive(&priv->sem);
+}
+
+static int ist415_close(FAR struct file *filep)
+{
+	printf("fd close \n");
```

Comment:
1. printf -> touchvdbg.
2. printing location should be moved to proper location. Because there is invalid return.
3. need to say which fd like `close touch fd %d`